### PR TITLE
Fix zero gradient for subtensor assignment.

### LINKED ISSUE #126: Gradient direct assignment is missing
When using direct assignment and the tensor being assigned is a function of parameters, those parameters don't seem to get a gradient i.e. `params.W1[i] = params.W2 * x` will result in a zero gradient for `W2`. The following code is a minimal test case:

```lua
grad = require 'autograd'
torch = require 'torch'

params={
  W=torch.range(0, 8):view(3, 3),
  storage=torch.zeros(3, 3)
}

function f(params, x)
  params.storage[2] = params.W * x
  return torch.mean(params.storage)
end

grad, _ = grad(f)(params, torch.ones(3))
print(grad.W)
```
The gradient of `W` here is zero, while it should be `torch.ones(3, 3) / 9`.

When I try to run the same thing with `{optimize=true}` I get an error:

```
/home/vanmerb/torch/install/bin/luajit: ...re/lua/5.1/autograd/runtime/codegen/backend/lua/init.lua:550: attempt to index local 'node' (a nil value)
stack traceback:
        ...re/lua/5.1/autograd/runtime/codegen/backend/lua/init.lua:550: in function 'addNodeTargets'
        ...re/lua/5.1/autograd/runtime/codegen/backend/lua/init.lua:572: in function 'generateCode'
        ...re/lua/5.1/autograd/runtime/codegen/backend/lua/init.lua:748: in function 'generateFn'
        .../install/share/lua/5.1/autograd/runtime/codegen/init.lua:140: in function <.../install/share/lua/5.1/autograd/runtime/codegen/init.lua:114>
        autograd_subtensor.lua:14: in main chunk
        [C]: in function 'dofile'
        ...merb/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
        [C]: at 0x00405e90
```

### PR DESCRIPTION
**Pull request by bartvm**
@alexbw Partially fixes #126 and adds a unit test. It gives the correct answer in direct mode now, but still crashes in optimized mode.

### CODE REVIEW

**Comment by alexbw** [comment]
Will merge when Assignment test passes on CI

**Comment by bartvm** [comment]
The test that I added fails on the current master as well, it's a bug somewhere in optimized mode. I'd have to get a bit more familiar with the details of optimized mode before I can fix that.

**Comment by bartvm** [comment]
Maybe someone with better knowledge of the optimized mode internals can understand what is going on. The problem seems to be [this line](https://github.com/twitter/torch-autograd/blob/0e7e28c2c51778d855bf5531a071e4df617cef85/src/runtime/codegen/Node.lua#L78) in `Node.lua` which specifically deals with `"Value.__internal_set`:

```lua
mutationFlow:alias(self.inputs[1], valueAlias)
```
Unlike `valueAlias = Value.from(outputs[1], Source.computed(self, 1))`, `self.inputs[1]` here is not transformed into a `Value` instance. [This line](https://github.com/twitter/torch-autograd/blob/master/src/runtime/codegen/backend/lua/init.lua#L572) in `runtime/codegen/backend/lua/init` takes issue with that:

```lua
aliasOp = graph.mutationFlow.history[i]
addNodeTargets(aliasOp.from.source.node, hazardNodes)
```
since `aliasOp.from.source` doesn't have a node, it's simply a table that looks like this:
```
from    {
  type : "tensor"
  source :
    {
      type : "table"
      parent :
        {
          type : "param"
          name : 1
        }
      key : "x"
    }
  raw : DoubleTensor - size: 10x10
}
```

**Comment by bartvm** [comment]
Because I need this for something else as well now, another stab at it. I'm using `torch.clone` to sidestep the issue of `params.x` not being a node (as was done in the other tests).

However, now the test for optimized mode fails for another reason: The gradient gets computed correctly the first time `grad(f)` is called, but when it is called a second time it overwrites the gradient in place and sets it to zero. This surfaced accidentally because `gradcheck` happens to call `grad(f)` twice to check for determinism. I made the check more explicit, but I haven't figured out how to fix it yet.

**Comment by bartvm** [comment]
@luketwitter @alexbw Although I can't figure out the current bug in optimized mode triggered by the new unit test I added, can I propose at least merging my changes to `src/gradfuns.lua`? That way the gradients for index assignment in direct mode aren't silently set to zero anymore, which is a really nasty bug because it can make models completely fail to train without giving a single error.

I made a new PR for this: #139
